Add compression controls

martinsumner opened the pull request:
Add configuration options to allow for a compression algorithm of `none` to disable compression altogether.  Also an option to change the point in the LSM tree when compression is applied.

martinsumner committed:
Add compression controls

Add configuration options to allow for a compression algorithm of `none` to disable compression altogether.  Also an option to change the point in the LSM tree when compression is applied.

ThomasArts commented on priv/leveled_multi.schema:
```diff
@@ -39,8 +39,8 @@
-  {default, lz4},
-  {datatype, atom},
+  {datatype, {enum, [native, lz4, none]}},
+  {default, native}
```
Sure to change the default here?

martinsumner replied:
It was a mistake.  The default for leveled backend was changed to native, but in that fix, forgot to change the multi-backend.  So this realigns the two again.

For riak, as it supports a HTTP API, there tends to be a lot of stuff that ends up being stored after base64 encoding.  In this case we get a 30% compression improvement native via lz4. The throughput improvement we've seen with lz4 due to efficient CPU has always been no more than about 3% except for contrived situations.

ThomasArts commented on include/leveled.hrl:
```diff
@@ -55,6 +55,7 @@
 -record(sst_options,
                         {press_method = native
                             :: leveled_sst:press_method(),
+                        press_level = 1 :: non_neg_integer(),
```
In leveled_bookie you define COMPRESSION_LEVEL to be 1.
I think, it would be nicer to define that default here and use it:
```suggestion
                        press_level = ?COMPRESSION_LEVEL :: non_neg_integer(),
```
Otherwise you encode the default in more than one place and that might be leading to future mistakes

martinsumner replied:
I agree.  There's quite a few other things where we end up with double-definitions of defaults, but I think it makes sense to correct them all as part of this commit.  In general all configurable items should have their defaults in `leveled.hrl`

ThomasArts commented on src/leveled_sst.erl:
```diff
@@ -78,6 +78,10 @@
 -define(COMPRESSION_LEVEL, 1).
+    % When using native compression - how hard should the compression code
+    % try to reduce the size of the compressed output. 1 Is to imply minimal
+    % effort, 6 is default:
```
but not the default of leveled, because here 1 is default

martinsumner replied:
Changed to COMPRESSION_FACTOR as this is different to the configurable COMPRESSION_LEVEL.  Will add reference to default being OTP not leveled.

ThomasArts reviewed (approved):
Looks good to me. Just double check you want to chaneg the default from lz4 to native at indicated place

martinsumner committed:
Handle configurable defaults consistently

Move them into leveled.hrl.  This forces double-definitions to be resolved.

There are some other constants in leveled_bookie that are relevant outside of leveled_bookie.  These are all now in the non-configurable startup defaults section.

martinsumner committed:
Clarify referred-to default is OTP not leveled
